### ISSUE TITLE
Dragonfruit and c#9-style top-level programs?

### ISSUE THREAD

**Issue by Trolldemorted** [dotnet/command-line-api#1071]
Dragonfruit is the only thing holding me back from switching to top-level programs in multiple projects. Do you have ideas how dragonfruit could work in one?

**Comment by KathleenDollard** [comment]
I'm working on where we go with the next generation of DragonFruit. In top level programs, I suspect there would be one line of code to get to your code. Would that be acceptable in a top level program in your view? Basically we need something to indicate where you are storing your CLI shape.

That work is on StarFruit2 in my repos, but it's currently a WIP, as I work out the structure of the generated classes, a decent approach to source generation (currently experimenting with a simple code DOM instead of helper methods) and the code needed at the entry point. Each of those is in draft, but I spent a good bit of time trying to make the helper methods for source gen work well. All three have to be working well for it to actually function.

Happy to have feedback on what you want from the next generation of DragonFruit. It would be great if you could rank the following and add more:

* Not having to learn much about System.CommandLine
* Ability to customize Options and Arguments (like setting a calculated default value)
* Subcommands for two layer and deeper CLIs
* Top level program support
* GenericHost
* Final validation/calculations before your code runs
* Super easy to get started (discovery on just namespace)
* Zero code (Just Main arguments (which is incompatible with top level programs))
* Almost no code

**Comment by Trolldemorted** [comment]
>I'm working on where we go with the next generation of DragonFruit. In top level programs, I suspect there would be one line of code to get to your code. Would that be acceptable in a top level program in your view? Basically we need something to indicate where you are storing your CLI shape.

Awesome to hear you are already on it! Since the declaration of the cli shape has to be somewhere, the top-level file would probably be best - new files are ugly, and the .csproj sounds like overkill.

I am a big fan of "almost no code", but it fear that does not go too well with multiple layer CLIs. My personal priorities would be final validation/calculations before your code runs, top level program support, and almost no code.

Since the requirements of top-level programs and old-style main functions diverge a lot, maybe you want to split top-level programm support into a separate package?

**Comment by heaths** [comment]
While I don't mind the rather verbose code required for System.CommandLine for dev tools or production apps, DragonFruit works great for customer samples I write often. It's concise and appears relatively unimportant - just want I want in a sample. Keeping that same concept would be nice. Do you have any prototype samples to share for feedback?

**Comment by chausner-audeering** [comment]
> DragonFruit works great for customer samples

I agree. In my opinion, DragonFruit really shines when you don't want to pollute your short program/sample with a lot of CLI parsing code but still want to provide basic functionality of command-line parsing and help.

Of course, DragonFruit only works for the simplest usecases at the moment. Two things I'd like to see added (ideally while keeping the simplicity):
* marking parameters as mandatory
* allowing parameters to have aliases/short and long forms (e.g. --global, -g)

I could imagine attributes could be useful to provide such additional information without requiring any additional code in the main function body.

**Comment by jboarman** [comment]
Is it conceivable to offer a model-based approach somewhat similar to how the CommandLineParser package handles this?  Perhaps we could place an attribute on the model (where only one/first model per assembly is used) so that we can use reflection to access the model?

https://github.com/commandlineparser/commandline/wiki#usage-attribute

**Comment by jonsequitur** [comment]
We're exploring updates to the more declarative DragonFruit approach. We plan to avoid runtime reflection by using source generators. Attributes will likely be part of the programming model, but not runtime reflection, which we're moving away from in order to make System.CommandLine faster and friendly to trimming.
